Extensions/TokenAcquisition.cs

```diff
@@ -0,0 +1,188 @@
+        /// <paramref name="user"/> parameter
+        /// </summary>
+        /// <param name="context">HttpContext (for instance of the controller)</param>
+        /// <param name="user">Account described by its claims</param>
+        /// <param name="scopes">Scopes to request for the downstream API to call</param>
+        /// <returns>An access token to call the downstream API characterized by its scopes, on behalf of the user</returns>
+        public async Task<string> GetAccessTokenOnBehalfOfUser(HttpContext context, ClaimsPrincipal user, string[] scopes)
+        {
+            // Use MSAL to get the right token to call the API
+            var credential = new ClientCredential(azureAdOptions.ClientSecret);
+            var request = context.Request;
+            var currentUri = UriHelper.BuildAbsolute(request.Scheme, request.Host, request.PathBase, request.Path);
+            Application = new ConfidentialClientApplication(azureAdOptions.ClientId, currentUri, new ClientCredential(azureAdOptions.ClientSecret), 
+                AuthPropertiesTokenCacheHelper.ForApiCalls(context, CookieAuthenticationDefaults.AuthenticationScheme), null);
+
+            string userObjectId = user.FindFirst("http://schemas.microsoft.com/identity/claims/objectidentifier").Value;
+            string tenantId = user.FindFirst("http://schemas.microsoft.com/identity/claims/tenantid").Value;
+            if (string.IsNullOrWhiteSpace(userObjectId))
+            {
+                // TODO: find a better typed exception
+                throw new Exception("Missing claim 'http://schemas.microsoft.com/identity/claims/objectidentifier'");
+            }
+
+            if (string.IsNullOrWhiteSpace(tenantId))
+            {
+                throw new Exception("Missing claim 'http://schemas.microsoft.com/identity/claims/tenantid'");
+            }
+            string accountId = userObjectId + "." + tenantId;
+
+            return await GetAccessTokenOnBehalfOfUser(accountId, scopes);
+        }
+
+        /// <summary>
+        /// Gets an access token for a downstream API on behalf of the user which account ID is passed as an argument
+        /// </summary>
+        /// <param name="accountIdentifier">User account identifier for which to acquire a token. 
+        /// See <see cref="Microsoft.Identity.Client.AccountId.Identifier"/></param>
+        /// <param name="scopes">Scopes for the downstream API to call</param>
+        public async Task<string> GetAccessTokenOnBehalfOfUser(string accountIdentifier, string[] scopes)
+        {
+            var accounts = (await Application.GetAccountsAsync());
+
+            string accessToken = null;
+            try
+            {
+                AuthenticationResult result = null;
+                IAccount account = await Application.GetAccountAsync(accountIdentifier);
+                result = await Application.AcquireTokenSilentAsync(scopes, account);
+                accessToken = result.AccessToken;
```

Comment:
could just return result.AccessToken and remove the variable - 'accessToken'

Reply:
Good point!